src/www/vpn_ipsec_pubkeys.php

```diff
@@ -0,0 +1,203 @@
+<?php
+
+/*
+ * Copyright (C) 2019 Pascal Mathis <mail@pascalmathis.com>
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES,
+ * INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY
+ * AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY,
+ * OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+require_once("guiconfig.inc");
+require_once("filter.inc");
+require_once("services.inc");
+require_once("interfaces.inc");
+require_once("plugins.inc.d/ipsec.inc");
+require_once('phpseclib/Math/BigInteger.php');
+require_once('phpseclib/Crypt/Hash.php');
+require_once('phpseclib/Crypt/RSA.php');
```

Comment:
is this needed or will the openssl functions do it as well?

Reply:
I have chosen phpseclib as the builtin OpenSSL functions are extremely restrictive about the supported key formats. As an example, OpenSSL only allows PKCS#1 keys (`-----BEGIN PUBLIC KEY-----`) and fails when parsing PKCS#8 keys (`-----BEGIN RSA PUBLIC KEY-----`). As both key formats are fairly common and Strongswan officially supports both of them as well, I did not want to manually restrict the plugin to a single format -or- mangle the key strings manually so that OpenSSL is satisified.

A short example is available here: https://repl.it/repls/GrownWarlikeQueryoptimizer

Reply:
In my opinion an automatic conversion to OpenSSL would make sense if it fails in first place. @fichtner What do you think?

Reply:
I gave this another thought and ended up rewriting parts of my implementation - aside from a few cleanups mentioned above, I moved all the heavy lifting of parsing, validating and sanitizing the RSA keys into a single function within `ipsec.inc` called `ipsec_parse_rsa_key($key_str, $key_type)`.

As every single documentation I could find for public key authentication actually says you should use the `ipsec pki` commands provided by Strongswan itself, I think accepting only X.509 keys is sufficient for 99.9% of use cases, as the Strongswan tool will always output the keys in said format. It will save us from additional complexity of mangling key formats (which tend to get complicated in specific situations) and I was able to fully drop the dependency on phpseclib, replacing it with native OpenSSL functions.

Both keys are now being properly sanitized when saved, so any unnecessary spaces and newlines get removed automatically. Additionally, the check if the public and private key belong to each other is now implemented by comparing the fingerprint, which is more efficient than signing and verifying a dummy message.

Last but not least I also changed the fingerprinting algorithm for the RSA keys to match the `keyid` outputted by `ipsec listall`, as it seemed like a more appropriate choice than the previous ssh-rsa based fingerprinting which does not show up in Strongswan at all.

Thank you for the good inputs and code review so far. I am fairly satisfied with the new implementation, any thoughts or good-to-go?